spt/features/ihud.cpp

```diff
@@ -0,0 +1,278 @@
+		}
+	}
+}
+
+void InputHud::PreHook()
+{
+	loadingSuccessful = !!(ORIG_DecodeUserCmdFromBuffer);
+}
+
+void InputHud::UnloadFeature() {}
+
+void InputHud::SetInputInfo(int button, Vector movement)
+{
+	if (awaitingFrameDraw)
+	{
+		buttonBits |= button;
+	}
+	else
+	{
+		buttonBits = button;
+	}
+	inputMovement = movement;
+	awaitingFrameDraw = true;
+}
+
+void InputHud::DrawInputHud()
+{
+	if (awaitingFrameDraw)
+	{
+		previousAng = currentAng;
+		float va[3];
+		EngineGetViewAngles(va);
+		currentAng = Vector(va[0], va[1], va[3]);
+		awaitingFrameDraw = false;
+	}
+	auto surface = spt_hud.surface;
+	auto scheme = spt_hud.scheme;
+	auto screen = spt_hud.screen;
+	ihudFont = scheme->GetFont(y_spt_ihud_font.GetString(), false);
+	const int IN_ATTACK = (1 << 0);
+	const int IN_JUMP = (1 << 1);
+	const int IN_DUCK = (1 << 2);
+	const int IN_FORWARD = (1 << 3);
+	const int IN_BACK = (1 << 4);
+	const int IN_USE = (1 << 5);
+	const int IN_MOVELEFT = (1 << 9);
+	const int IN_MOVERIGHT = (1 << 10);
+	const int IN_ATTACK2 = (1 << 11);
+
+	int xOffset = y_spt_ihud_x.GetInt();
+	int yOffset = y_spt_ihud_y.GetInt();
+	int gridSize = y_spt_ihud_grid_size.GetInt();
+	int padding = y_spt_ihud_grid_padding.GetInt();
+
+	Color buttonColor;
+	Color shadowColor;
+	Color buttonFontColor;
+	Color shadowFontColor;
+
+#define GET_COLOR(cmd, color) \
+	{ \
+		static int r = 0, g = 0, b = 0, a = 0; \
+		if (strcmp("", cmd.GetString()) != 0) \
+		{ \
+			sscanf(cmd.GetString(), "%d %d %d %d", &r, &g, &b, &a); \
+		} \
+		color = Color(r, g, b, a); \
+	}
+
+	GET_COLOR(y_spt_ihud_button_color, buttonColor);
+	GET_COLOR(y_spt_ihud_shadow_color, shadowColor);
+	GET_COLOR(y_spt_ihud_font_color, buttonFontColor);
+	GET_COLOR(y_spt_ihud_shadow_font_color, shadowFontColor);
+
+#define DRAW_BUTTON(button, col, row, size, colSize, rowSize, text) \
+	{ \
+		Color btnColor = (buttonBits & (button)) ? buttonColor : shadowColor; \
+		Color fontColor = (buttonBits & (button)) ? buttonFontColor : shadowFontColor; \
+		DrawRectAndCenterTxt(btnColor, \
+		                     xOffset + col * (size + padding), \
+		                     yOffset + row * (size + padding), \
+		                     xOffset + (col + colSize) * (size + padding) - padding, \
+		                     yOffset + (row + rowSize) * (size + padding) - padding, \
+		                     ihudFont, \
+		                     fontColor, \
+		                     surface, \
+		                     text); \
+	}
+
+	if (y_spt_ihud.GetInt() == 2)
+	{
+		gridSize /= 2;
+
+		// Angle
+		Vector ang = currentAng - previousAng;
+		while (ang.x < -180.0f)
+			ang.x += 360.0f;
+		if (ang.x > 180.0f)
+			ang.x -= 360.0f;
+		float len = ang.Length();
+		if (len > 0)
+		{
+			ang /= len;
+			ang *= pow(len / 180.0f, 0.2);
+			ang.y *= -1;
+		}
+
+		// Movement
+		Vector movement = inputMovement;
+		int movementSpeed = movement.Length();
+		int maxSpeed = movementSpeed > 150 ? 450 : 150; // For Portal TAS
```

Comment:
Include header ent_utils.hpp and use `utils::GetProperty<float>(0, "m_flMaxSpeed");` instead of hardcoding the maxspeed.